pkg/ddl/affinity.go

```diff
@@ -0,0 +1,197 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"fmt"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/ddl/logutil"
+	"github.com/pingcap/tidb/pkg/domain/infosync"
+	"github.com/pingcap/tidb/pkg/meta/model"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/tablecodec"
+	tikv "github.com/tikv/client-go/v2/tikv"
+	pdhttp "github.com/tikv/pd/client/http"
+	"go.uber.org/zap"
+)
+
+// GetTableAffinityGroupID returns the affinity group ID for a table.
+// Format: "_tidb_t_{tableID}"
+func GetTableAffinityGroupID(tableID int64) string {
+	return fmt.Sprintf("_tidb_t_%d", tableID)
+}
+
+// GetPartitionAffinityGroupID returns the affinity group ID for a partition.
+// Format: "_tidb_pt_{tableID}_p{partitionID}"
+func GetPartitionAffinityGroupID(tableID, partitionID int64) string {
+	return fmt.Sprintf("_tidb_pt_%d_p%d", tableID, partitionID)
+}
+
+func buildAffinityGroupKeyRange(codec tikv.Codec, physicalID int64) pdhttp.AffinityGroupKeyRange {
+	startKey := tablecodec.EncodeTablePrefix(physicalID)
+	endKey := tablecodec.EncodeTablePrefix(physicalID + 1)
+	if codec != nil {
+		startKey, endKey = codec.EncodeRegionRange(startKey, endKey)
+	}
+	return pdhttp.AffinityGroupKeyRange{
+		StartKey: startKey,
+		EndKey:   endKey,
+	}
+}
+
+// buildAffinityGroupDefinitions constructs affinity group definitions based on table's affinity configuration.
+// It generates affinity group IDs in two different formats depending on the affinity level:
+//   - Table-level affinity: "_tidb_t_{tableID}" - one group for the entire table
+//   - Partition-level affinity: "_tidb_pt_{tableID}_p{partitionID}" - one group per partition
+func buildAffinityGroupDefinitions(codec tikv.Codec, tblInfo *model.TableInfo, partitionDefs []model.PartitionDefinition) (map[string][]pdhttp.AffinityGroupKeyRange, error) {
+	if tblInfo == nil || tblInfo.Affinity == nil {
+		return nil, nil
+	}
+
+	switch tblInfo.Affinity.Level {
+	case ast.TableAffinityLevelTable:
+		groupID := GetTableAffinityGroupID(tblInfo.ID)
+		return map[string][]pdhttp.AffinityGroupKeyRange{
+			groupID: {buildAffinityGroupKeyRange(codec, tblInfo.ID)},
+		}, nil
+	case ast.TableAffinityLevelPartition:
+		definitions := partitionDefs
+		if definitions == nil && tblInfo.Partition != nil {
+			definitions = tblInfo.Partition.Definitions
+		}
+		if len(definitions) == 0 {
+			return nil, errors.Errorf("partition affinity requires partition definitions for table %s (ID: %d), table metadata may be corrupted", tblInfo.Name.O, tblInfo.ID)
+		}
+
+		groups := make(map[string][]pdhttp.AffinityGroupKeyRange, len(definitions))
+		for _, def := range definitions {
+			groupID := GetPartitionAffinityGroupID(tblInfo.ID, def.ID)
+			groups[groupID] = []pdhttp.AffinityGroupKeyRange{buildAffinityGroupKeyRange(codec, def.ID)}
```

Comment:
So, it's maintaining a group ID => key range mapping, and key range is calculated from table / partition ID.
I wonder why not just groupID => table / partition ID?